mcstatus/motd.py

```diff
@@ -0,0 +1,549 @@
+    MinecraftColor.DARK_AQUA: ("#00AAAA", "#002A2A"),
+    MinecraftColor.DARK_RED: ("#AA0000", "#AA0000"),
+    MinecraftColor.DARK_PURPLE: ("#AA00AA", "#2A002A"),
+    MinecraftColor.GOLD: ("#FFAA00", "#402A00"),
+    MinecraftColor.GRAY: ("#AAAAAA", "#2A2A2A"),
+    MinecraftColor.DARK_GRAY: ("#555555", "#151515"),
+    MinecraftColor.BLUE: ("#5555FF", "#15153F"),
+    MinecraftColor.GREEN: ("#55FF55", "#153F15"),
+    MinecraftColor.AQUA: ("#55FFFF", "#153F3F"),
+    MinecraftColor.RED: ("#FF5555", "#3F1515"),
+    MinecraftColor.LIGHT_PURPLE: ("#FF55FF", "#3F153F"),
+    MinecraftColor.YELLOW: ("#FFFF55", "#3F3F15"),
+    MinecraftColor.WHITE: ("#FFFFFF", "#3F3F3F"),
+    MinecraftColor.MINECOIN_GOLD: ("#DDD605", "#373501"),
+}
+
+MINECRAFT_COLOR_TO_HEX_JAVA = MINECRAFT_COLOR_TO_HEX_BEDROCK.copy()
+MINECRAFT_COLOR_TO_HEX_JAVA[MinecraftColor.GOLD] = ("#FFAA00", "#2A2A00")
+
+MINECRAFT_COLOR_TO_ANSI = {
+    MinecraftColor.BLACK: "30",
+    MinecraftColor.DARK_BLUE: "34",
+    MinecraftColor.DARK_GREEN: "32",
+    MinecraftColor.DARK_AQUA: "36",
+    MinecraftColor.DARK_RED: "31",
+    MinecraftColor.DARK_PURPLE: "35",
+    MinecraftColor.GOLD: "33",
+    MinecraftColor.GRAY: "37",
+    MinecraftColor.DARK_GRAY: "90",
+    MinecraftColor.BLUE: "94",
+    MinecraftColor.GREEN: "92",
+    MinecraftColor.AQUA: "96",
+    MinecraftColor.RED: "91",
+    MinecraftColor.LIGHT_PURPLE: "95",
+    MinecraftColor.YELLOW: "93",
+    MinecraftColor.WHITE: "97",
+    MinecraftColor.MINECOIN_GOLD: "33",
+}
+
+
+@dataclass
+class WebColor:
+    """Raw HTML color from MOTD.
+
+    Can be found in MOTD when someone uses gradient.
+    """
+
+    hex: str
+    rgb: tuple[int, int, int]
+
+    @classmethod
+    @overload
+    def parse(cls, *, hex: str, rgb: Optional[tuple[int, int, int]] = None) -> Self:
+        ...
+
+    @classmethod
+    @overload
+    def parse(cls, *, rgb: tuple[int, int, int], hex: Optional[str] = None) -> Self:
+        ...
+
+    @classmethod
+    def parse(cls, *, hex: Optional[str] = None, rgb: Optional[tuple[int, int, int]] = None) -> Self:
+        """Parse hex or RGB color, from hex or RGB (one of them can be ``None``).
+
+        Ensures that ``WebColor.hex`` starts with '#', and generate RGB color for it,
+        if ``WebColor.rgb`` is None (default value). Or generate hex value from RGB, if
+        ``WebColor.hex`` is None (also default value).
+
+        :raises ValueError: When hex color isn't correct, so we can't generate RGB for it.
+        :raises TypeError: When hex and RGB colors is ``None``.
+        :returns: New ``WebColor`` instance.
+        """
+        if hex is None and rgb is None:
+            raise TypeError("`hex` and `rgb` parameters is None, we need one of them.")
+
+        if hex is None and rgb is not None:
+            if not all(256 > e > -1 for e in rgb):
+                raise ValueError("One of RGB colors is less than zero or more than 255: " + str(rgb))
+            hex = "#%02x%02x%02x" % rgb
```

Comment:
With cases like these, 2 class methods should be preferred over one cluttered method with overloads. Even if the hex method just converts the hex to rgb and calls the rgb one (or vice-versa).

Also, instead of doing `"#%02x%02x%02x" % rgb` we can just use format strings like: `hex = f"#{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}"`

```suggestion
    @classmethod
    def from_hex(cls, hex: str) -> Self:
        hex = hex.removeprefix("#")

        try:
            rgb = tuple(int(hex[i : i + 1], 16) for i in (0, 2, 4))
        except ValueError:
            raise ValueError(f"Failed to parse given hex color: {'#' + hex!r}")
        return cls.from_rgb(rgb)

    @classmethod
    def from_rgb(cls, rgb: tuple[int, int, int]) -> Self:
        if not all(255 >= e >= 0 for e in rgb):
            raise ValueError("Got RGB color out of it's 8-bit range (0-255)")

        hex = f"#{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}".upper()
        return cls(hex, rgb)
```

Reply:
Isn't `"#%02x%02x%02x" % rgb` more readable than `f"#{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}"`? I think no, so why should we choose second variant?

Everything others fixed in [this force-push](https://github.com/py-mine/mcstatus/compare/4596d541efe9be97dd7f92fb059e47784572e4cf..8b2935c714ffd2ac802fe9538a10dd356d7ae991), except for `rgb` generation in `from_hex` where you used `i : i + 1` instead of `+ 2` in my implementation (this change just breaks tests) and uppercase hex in `from_rgb`, as it is an additional useless operation.

Reply:
> Isn't "#%02x%02x%02x" % rgb more readable than f"#{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}"? I think no, so why should we choose second variant?

Another variant would be to use `"{:02x}{:02x}{:02x}".format(*rgb)`, or `{r:02x}{g:02x}{b:02x}`, with a line above doing `r, g, b = rgb` but I'm against using the % string formatting operator here, as that's a very outdated way to format strings, and some people aren't even familiar with it in modern python code. It's also a feature that's likely to get deprecated in comparison to `str.format`.

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/ffe1cc869956fc92fc7980d4c8a265d23adfa1b6..ae2dd4eff4d605ae236fba6a2f87e783aa51de2e#diff-c462c16c38f067b712f30e9e5d65eda0a525e97348b12492eacf676da5656ed6L173-R173).